File: 백준/Gold/13144. List of Unique Numbers/List of Unique Numbers.java
import java.io.*;
import java.util.*;

public class Main{

    // 필요변수
    static int N;
    static int[] nums;
    static int[] cnt = new int[100000 + 1];

    // 입력
    static void input() throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        N = Integer.parseInt(br.readLine());
        nums = new int[N + 1];

        st = new StringTokenizer(br.readLine());
        for (int i = 1; i <= N; i++) {
            nums[i] = Integer.parseInt(st.nextToken());
        }

        br.close();
    }

    // 풀이
    static void pro() {
        int R = 0;
        long ans = 0;

        for (int L = 1; L <= N + 1; L++) {
            cnt[nums[L - 1]] = 0;

            while (R + 1 <= N && cnt[nums[R + 1]] == 0) {
                cnt[nums[++R]] = 1;
            }

            ans += R - L + 1;
        }

        System.out.println(ans);
        
    }
    public static void main(String[] args) throws Exception{
        input();
        pro();
    }
}

File: 백준/Gold/13144. List of Unique Numbers/List of Unique Numbers.cc
#include <iostream>
#include <vector>

using namespace std;

// 변수
int N;
vector<int> v;
int nums[100000 + 1];

void input() {
    cin >> N;

    int temp;
    v.push_back(0);
    for (int i = 0; i < N; i++) {
        cin >> temp;
        v.push_back(temp);
    }
}

void pro() {
    int R = 0;
    long long cnt = 0;

    for (int L = 1; L <= N; L++) {
        // L - 1 제거
        nums[v[L - 1]] = 0;

        // 같은 수 나올 때까지 R 증가
        while (R + 1 <= N && nums[v[R + 1]] == 0) {
            nums[v[++R]]++;
        }

        cnt += R - L + 1;
    }

    // 출력
    cout << cnt;
}

int main() {
    // 최적화
    ios::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    // 입력
    input();

    // 실행
    pro();
}